server/src/main/kotlin/com/trib3/server/graphql/GraphQLWebSocketSubscriber.kt

```diff
@@ -0,0 +1,289 @@
+     * the connection and start a keepalive timer.
+     */
+    private fun handleConnectionInit(message: OperationMessage<*>) {
+        if (keepAlive == null) {
+            adapter.sendMessage(OperationType.GQL_CONNECTION_ACK, message.id)
+            adapter.sendMessage(OperationType.GQL_CONNECTION_KEEP_ALIVE, message.id)
+            val newKeepAlive = Flowable.interval(
+                graphQLConfig.keepAliveIntervalSeconds,
+                TimeUnit.SECONDS,
+                keepAliveScheduler
+            ).subscribe {
+                log.trace("WebSocket connection keepalive ping $it")
+                adapter.emitter.onNext(
+                    OperationMessage(
+                        OperationType.GQL_CONNECTION_KEEP_ALIVE,
+                        message.id
+                    )
+                )
+            }
+            keepAlive = newKeepAlive
+            disposables.add(newKeepAlive)
+        } else {
+            adapter.sendMessage(OperationType.GQL_CONNECTION_ERROR, message.id, "Already connected!")
+        }
+    }
+
+    /**
+     * Process an [OperationType.GQL_CONNECTION_TERMINATE] message.  Cancelling the subscription
+     * will dispose of any downstream subscriptions and close the WebSocket (see [afterSubscribed])
+     */
+    private fun handleConnectionTerminate(message: OperationMessage<*>) {
+        log.info("WebSocket connection termination requested by message ${message.id}!")
+        subscription.cancel()
+    }
+
+    /**
+     * Process an [OperationType.GQL_START] message.  Executes the graphQL query in a downstream
+     * subscriber which will asynchronously emit new events as the query returns data or errors.
+     * Tracks running queries by client specified id, and only allows one running query for a given id.
+     */
+    private fun handleQueryStart(message: OperationMessage<*>) {
+        val messageId = message.id
+        check(messageId != null) {
+            "Must pass a message id to start a query"
+        }
+        check(!queries.containsKey(messageId)) {
+            "Query with id $messageId already running!"
+        }
+        check(message.payload is GraphRequest) {
+            "Invalid payload for query"
+        }
+        check(graphQL != null) {
+            "graphQL not configured!"
+        }
+        val payload = message.payload as GraphRequest
+        val executionQuery = ExecutionInput.newExecutionInput()
+            .query(payload.query)
+            .variables(payload.variables ?: mapOf())
+            .operationName(payload.operationName)
+            .build()
+        val flowable = Flowable.defer {
+            RequestIdFilter.withRequestId(messageId) {
+                val result = graphQL.execute(executionQuery)
+                // check if the graphQL result is a Publisher
+                val publisherData = try {
+                    result.getData<Publisher<ExecutionResult>>()
+                } catch (e: Exception) {
+                    null
```

Comment:
log something here?

Reply:
that'll happen every time we do a regular `query` over the websocket, which doesn't seem worthy of a log.  i wish the ExecutionResult exposed a better way of introspecting its data, but this is the best i've got....